lib/Dialect/TritonGPU/Transforms/RemoveLayoutConversions.cpp

```diff
@@ -101,29 +101,59 @@ class SimplifyReduceCvt : public mlir::RewritePattern {
     auto convert = llvm::cast<triton::gpu::ConvertLayoutOp>(op);
     triton::ReduceOp reduce;
     for (auto &use : convert.getResult().getUses()) {
-      auto owner = use.getOwner();
-      if (llvm::isa_and_nonnull<triton::ReduceOp>(owner)) {
-        reduce = llvm::cast<triton::ReduceOp>(owner);
-        break;
+      auto owner = llvm::dyn_cast<triton::ReduceOp>(use.getOwner());
+      if (!owner) {
+        continue;
+      }
+
+      // TODO: This always takes layout from the first argument which
+      // is fine for argmin/argmax but may not be optimal generally
+      if (convert.getResult() != owner.getOperands()[0]) {
+        continue;
       }
+      reduce = owner;
+      break;
     }
     if (!reduce)
       return mlir::failure();
-    // this may generate unsupported conversions in the LLVM codegen
-    if (convert.getOperand()
-            .getType()
-            .cast<RankedTensorType>()
-            .getEncoding()
-            .isa<triton::gpu::MmaEncodingAttr>())
-      return mlir::failure();
+
+    SmallVector<Value> newOperands = reduce.getOperands();
+
+    newOperands[0] = convert.getOperand();
+    auto newEncoding =
+        newOperands[0].getType().cast<RankedTensorType>().getEncoding();
+
+    if (!newEncoding.isa<triton::gpu::BlockedEncodingAttr>()) {
```

Comment:
We don't enforce the BlockingEncoding requirement. The old code just says we don't push convert_layout with `MmaEncodingAttr` across the reduce op.

Reply:
Do you mean that it's okay for reduce to fail to compile if it ends up being called with an `MmaEncodingAttr`? Because the reduction does strictly require `BlockedEncodingAttr` as you can see with this line in the existing code:
```
   auto srcLayout = srcTy.getEncoding().cast<BlockedEncodingAttr>();
```

Reply:
No, I meant we do want to support `MmaEncodingAttr`. 

> auto srcLayout = srcTy.getEncoding().cast<BlockedEncodingAttr>();

This is buggy actually...We are working on a fix.